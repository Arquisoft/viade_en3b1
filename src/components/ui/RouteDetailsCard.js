import React from 'react';
import { makeStyles } from '@material-ui/core/styles';
import Card from '@material-ui/core/Card';
import CardContent from '@material-ui/core/CardContent';
import Typography from '@material-ui/core/Typography';
import Grid from '@material-ui/core/Grid';
import DetailsMap from '../map/DetailsMap.js';

const useStyles = makeStyles((theme) => ({
    root: {
        flexGrow: 1,
    },
}));

export default function RouteDetails(props) {
    const classes = useStyles();

    return (
        <Grid container className={classes.root} spacing={1}>
            <Grid item xs={6}>
                <Card elevation={5} className={classes.root}>
                    <CardContent>
                        <Typography variant="h5" component="h2">
                            {props.route.getName()}
                        </Typography>

                        <DetailsMap route={props.route}></DetailsMap>
                    </CardContent>
                </Card>
            </Grid>
            <Grid item xs={6}>
                <Grid container direction="column" spacing={1} alignItems="stretch" justify="flex-start">
                    <Grid item xs={12}>
                        <Card elevation={5} className={classes.root}>
                            <CardContent>
                                <Typography className={classes.pos} >
                                    {/* Date: {props.route.getDate()} */}
                                    Date: 22/3/2020
                                </Typography>
                            </CardContent>
                        </Card>
                    </Grid>
                    <Grid item xs={12}>
                        <Card elevation={5} className={classes.root}>
                            <CardContent>
                                <Typography className={classes.pos} >
                                    {/* Distance: {props.route.getTotalDistance()} */}
                                    Distance: 3.63km
                                </Typography>
                            </CardContent>
                        </Card>
                    </Grid>
                    <Grid item xs={12}>
                        <Card elevation={5} className={classes.root}>
                            <CardContent>
                                <Typography className={classes.pos} >
                                    {/* Time: {props.route.getTime()} */}
                                    Time: 42 min
                                </Typography>
                            </CardContent>
                        </Card>
                    </Grid>
                    <Grid item xs={12}>
                        <Card elevation={5} className={classes.root}>
                            <CardContent>
                                <Typography className={classes.pos} >
                                    Description:
                                </Typography>
                                <Typography className={classes.pos} >
                                    {props.route.getDescription()}
                                </Typography>
                            </CardContent>
                        </Card>
                    </Grid>
                    {/* <Grid item xs={12}>
                        <Card elevation={5} className={classes.root}>
                            <CardContent>
                                <Typography className={classes.pos} >
                                    Coments:
                                </Typography>
                                {
                                    Array.isArray(props.route.getComments()) && //is empty??
                                    props.route.getComments().map((comment) =>
                                        (
                                            <Typography className={classes.pos} >
                                                {comment}
                                            </Typography>
                                        )
                                    )
                                }
                            </CardContent>
                        </Card>
                    </Grid> */}
                </Grid>
            </Grid>
        </Grid>
    );
}