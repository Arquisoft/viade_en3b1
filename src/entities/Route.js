import { v4 as uuid } from "uuid";
import RouteElement from "./RouteElement";

class Route {

    /**
     * Constructor for a new Route. 
     * An id will be automatically generated (uuid v4).
     * TotalDistance will also be automatically generated.
     * 
     * RouteElements have name, latitude, longitude and elevation.
     * 
     * @param {String} name Name of the route.
     * @param {Date} date Date of the route.
     * @param {String} description Description of the route.
     * @param {Array} routeElements List of RouteElement objects, which are the trackpoints.
     * @param {Array} comments List of comments
     * @param {Array<String>} media List of urls of media (on the POD)
     */
    constructor(name, description, routeElements, comments, media, date) {
        this.id = uuid().toString();
        this.name = name;
        this.date = date;
        this.description = description;

        this.media = media;
        this.totalDistance = null;

        if (comments === null) {
            this.comments = [];
        } else {
            this.comments = comments;
        }

        if (routeElements[0] instanceof RouteElement) {
            this.routeElements = routeElements;
        } else {
            this.routeElements = generateRouteElements(routeElements);
        }
    }

    getId() {
        return this.id;
    }

    getName() {
        return this.name;
    }

    getDate() {
        return this.date;
    }

    getTime() {
        return this.time;
    }

    getDescription() {
        return this.description;
    }

    getRouteElements() {
        return this.routeElements;
    }

    getTotalDistance() {
        var total = 0;
        var aux = 0;

        for (let i = 1; i < this.routeElements.length; i++) {
            aux = this.routeElements[i].getLatitude() - this.routeElements[i - 1].getLatitude();
            total += this.addDistance(aux);
            aux = this.routeElements[i].getLongitude() - this.routeElements[i - 1].getLongitude();
            total += this.addDistance(aux);
            aux = this.routeElements[i].getElevation() - this.routeElements[i - 1].getElevation();
            total += this.addDistance(aux);
        }

        return total;
    }

    addDistance(aux) {
        if (aux < 0) {
            return -aux;
        } else {
            return aux;
        }
    }

    getComments() {
        return this.comments;
    }

    getMedia() {
        return this.media;
    }

    addMedia(media) {
        this.media.push(media);
    }

    addRouteElement(routeElement) {
        this.routeElements.push(routeElement);
    }

    getJsonLD() {
        let routePointsJson = [];
        this.routeElements.forEach((p) => routePointsJson.push(p.toJsonLatLng()));
        return JSON.stringify(
            {
                "@context": {
                    "@version": 1.1,
                    "comments": {
                        "@container": "@list",
                        "@id": "viade:comments"
                    },
                    "description": {
                        "@id": "schema:description",
                        "@type": "xs:string"
                    },
                    "media": {
                        "@container": "@list",
                        "@id": "viade:media"
                    },
                    "name": {
                        "@id": "schema:name",
                        "@type": "xs:string"
                    },
                    "points": {
                        "@container": "@list",
                        "@id": "viade:points"
                    },
                    "latitude": {
                        "@id": "schema:latitude",
                        "@type": "xs:double"
                    },
                    "longitude": {
                        "@id": "schema:longitude",
                        "@type": "xs:double"
                    },
                    "elevation": {
                        "@id": "schema:elevation",
                        "@type": "xsd:double"
                    },
                    "author": {
                        "@id": "schema:author",
                        "@type": "@id"
                    },
                    "rdf": "http://www.w3.org/1999/02/22-rdf-syntax-ns#",
                    "rdfs": "http://www.w3.org/2000/01/rdf-schema#",
                    "schema": "http://schema.org/",
                    "viade": "http://arquisoft.github.io/viadeSpec/",
                    "xsd": "http://www.w3.org/2001/XMLSchema#"
                },
                "name": this.name,
                "author": this.author,
                "description": this.description,
                "comments": this.comments,
                "media": this.media,
                "points": routePointsJson
            }
        );
    }

}

/**
 * Method that receives a list of points and returns
 * a list of RouteElements
 * @param {Array} points List of points containing name, latitude, longitude, elevation. 
 */
function generateRouteElements(points) {
    let routeElements = [];
    points.forEach((p) => routeElements.push(new RouteElement(p.lat, p.lng)));
    return routeElements;
}

export default Route;