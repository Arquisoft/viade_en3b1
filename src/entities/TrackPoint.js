class TrackPoint {

    /**
     * 
     * @param {Number} latitude 
     * @param {Number} longitude 
     * @param {String} name 
     * @param {Number} elevation 
     */
    constructor(latitude, longitude, name, elevation) {

        this.latitude = latitude;
        this.longitude = longitude;

        if (!name) {
            this.name = "";
        } else {
            this.name = name;
        }

        if(!elevation){
            this.elevation = "";
        } else {
            this.elevation = elevation;
        }
    }

    getName() {
        return this.name;
    }

    getLatitude() {
        return this.latitude;
    }

    getElevation() {
        return this.elevation;
    }

    getLongitude() {
        return this.longitude;
    }

    setElevation(num) {
        this.elevation = num;
    }

    toJson() {
        return {
            "latitude": this.latitude,
            "longitude": this.longitude,
            "elevation": this.elevation,
        };
    }
}

export default TrackPoint;