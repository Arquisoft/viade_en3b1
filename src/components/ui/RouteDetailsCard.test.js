import RouteDetailsCard from "./RouteDetailsCard";
import Route from '../../entities/Route';
import TrackPoint from '../../entities/TrackPoint';

var route1 = new Route("Route1", "Description 1", [new TrackPoint(1,1), new TrackPoint(2,1)], null, null, null);

test('Define RouteDetailsCard', () => {
    expect(RouteDetailsCard).toBeDefined();
});

test('Renders correctly', () => {
    const wrapper = render(<RouteDetailsCard route={route1} />);
    expect(wrapper).toMatchSnapshot();
});