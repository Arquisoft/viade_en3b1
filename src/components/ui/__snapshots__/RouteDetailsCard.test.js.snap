// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders correctly 1`] = `
<div>
  <a
    aria-disabled="false"
    aria-label="back"
    class="MuiButtonBase-root MuiFab-root makeStyles-backBtn-15 MuiFab-sizeSmall"
    href="#/dashboard"
    tabindex="0"
    title="Back"
  >
    <span
      class="MuiFab-label"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20v-2z"
        />
      </svg>
    </span>
  </a>
  <div
    class="MuiGrid-root makeStyles-root-1 MuiGrid-container MuiGrid-spacing-xs-3"
  >
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true MuiGrid-grid-sm-6"
    >
      <div
        class="MuiGrid-root makeStyles-title-8 MuiGrid-item"
      >
        <div
          class="MuiBox-root MuiBox-root-153"
        >
          Route1
        </div>
        <h6
          class="MuiTypography-root MuiTypography-h6 MuiTypography-colorTextSecondary"
        >
          by you
        </h6>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item"
      >
        <div
          class="MuiGrid-root makeStyles-mapTitle-3 MuiGrid-container MuiGrid-item"
        >
          <img
            alt="Map logo"
            class="makeStyles-maplogo-2"
            src="map.svg"
          />
          <div
            class="MuiBox-root MuiBox-root-184"
          >
            Map
          </div>
        </div>
        <div
          class="MuiPaper-root MuiCard-root makeStyles-map-4 MuiPaper-outlined MuiPaper-rounded"
        >
          <div
            id="mapdetails"
          />
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item"
      >
        <div
          class="MuiGrid-root makeStyles-elevationTitle-7 MuiGrid-container MuiGrid-item"
        >
          <img
            alt="Elevation logo"
            class="makeStyles-elevationlogo-6"
            src="mountain.svg"
          />
          <div
            class="MuiBox-root MuiBox-root-214"
          >
            Elevation
          </div>
        </div>
        <div
          class="MuiGrid-root makeStyles-chart-5 MuiGrid-item"
        >
          <canvas
            height="100"
            width="300"
          />
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true MuiGrid-grid-sm-4"
    >
      <div
        class="MuiGrid-root MuiGrid-item"
      >
        <div
          class="MuiGrid-root makeStyles-carouselTitle-10 MuiGrid-container MuiGrid-item"
        >
          <img
            alt="Camera logo"
            class="makeStyles-cameralogo-9"
            src="camera.svg"
          />
          <div
            class="MuiBox-root MuiBox-root-215"
          >
            Gallery
          </div>
        </div>
        <div
          class="MuiPaper-root MuiCard-root makeStyles-carousel-11 MuiPaper-rounded"
        >
          <div
            class="awssld awssld--organic-arrows"
          >
            <div
              class="awssld__wrapper"
            >
              <div
                class="awssld__container"
              >
                <div
                  class="awssld__box"
                >
                  <div
                    class="awssld__content"
                  >
                    <img
                      src="city_1.jpg"
                    />
                  </div>
                </div>
                <div
                  class="awssld__box"
                />
              </div>
              <div
                class="awssld__controls awssld__controls--hidden"
              >
                <button
                  aria-label="next"
                  class="awssld__next"
                >
                  <span
                    class="awssld__controls__arrow-right"
                  />
                </button>
                <button
                  aria-label="previous"
                  class="awssld__prev"
                >
                  <span
                    class="awssld__controls__arrow-left"
                  />
                </button>
              </div>
            </div>
            <nav
              class="awssld__bullets"
            >
              <button
                class="awssld__bullets--active"
                data-index="0"
              >
                0
              </button>
              <button
                data-index="1"
              >
                1
              </button>
            </nav>
          </div>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item"
      >
        <div
          class="MuiGrid-root makeStyles-dataTitle-14 MuiGrid-container MuiGrid-item"
        >
          <img
            alt="Map logo"
            class="makeStyles-datalogo-13"
            src="arrows.svg"
          />
          <div
            class="MuiBox-root MuiBox-root-216"
          >
            Info
          </div>
        </div>
        <div
          class="MuiPaper-root MuiCard-root makeStyles-data-12 MuiPaper-outlined MuiPaper-rounded"
        >
          <div
            class="MuiCardContent-root"
          >
            <div
              class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3"
            >
              <div
                class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3 MuiGrid-item"
              >
                <div
                  class="MuiGrid-root MuiGrid-item"
                >
                  <div
                    class="MuiBox-root MuiBox-root-218"
                  >
                    Date
                  </div>
                  <div
                    class="MuiBox-root MuiBox-root-219"
                  >
                    --
                  </div>
                </div>
                <div
                  class="MuiGrid-root MuiGrid-item"
                >
                  <div
                    class="MuiBox-root MuiBox-root-220"
                  >
                    Distance
                  </div>
                  <div
                    class="MuiBox-root MuiBox-root-221"
                  >
                    111194
                  </div>
                </div>
              </div>
              <div
                class="MuiGrid-root MuiGrid-item"
              >
                <div
                  class="MuiGrid-root MuiGrid-item"
                >
                  <div
                    class="MuiBox-root MuiBox-root-222"
                  >
                    Description
                  </div>
                  <div
                    class="MuiBox-root MuiBox-root-223"
                  >
                    Description 1
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Renders null 1`] = `null`;
